### PR TITLE
[INFINITY-1609] [INFINITY-1640] Additional commands for update

### PR DESCRIPTION
**Pull request by ssk2**
I've essentially just created some new commands for `dcos <service> update` that re-use the handlers for the equivalent commands in `dcos <service> plan`.

Several of these were renamed (but have aliases for the previous commands):
+ continue -> resume
+ interrupt -> pause
+ restart -> force-restart

This PR also includes the `dcos <service> update package-versions` command. 

Other changes made:
* I reordered commands and methods in `plan.go` to be in alphabetical order.
* Moved two tests from `update_test.go` to `plan_test.go`.
* Moved printing logic out of `response.go` into `print.go`.
* Changed `plan_test.go` into a TestSuite that will allow us to mock out responses for new tests.
* Fixed all the warnings generated by `golint` and unexported a number of commands that didn't need to be exported.
* Changed the methods in http.go to return `[]byte` instead of Response (this keeps all the code that operates on Response structs in the same library at least).
* Changed some of the methods to return errors instead of printing directly to stdout. I'm hoping this is a first step to eventually getting to a place where we can use Kingpin's error handling logic.
* Fixed a bug in the Plans API where it would only 404 for the first request that came in (and would stacktrace thereafter).
* Added some extra logic to the Plans API to return a 208 if the user has already submitted that request, or it was unable to be fulfilled because the plan had already completed.

What remains to be done:
- [x] Check that output is what it should be
- [x] Validate that this layout is acceptable to @keithchambers (I worry that having redundant options in the help output might be confusing).
- [x] Rebase on top of cli-plan-status when that is ready for merge
- [x] Integration tests for these commands
- [x] Test package-versions on a live cluster (blocked on https://github.com/dcos/dcos-cli/pull/1001).
- [ ] Better error catching for Cosmos errors.
- [x] Rollback

Docs updates will be done as part of #1053.

```sh
➜  helloworld git:(ssk2/additional-commands) ✗ dcos hello-world
usage: dcos hello-world [<flags>] <command> [<args> ...]

Flags:
  -h, --help                Show context-sensitive help (also try --help-long
                            and --help-man).
  -v, --verbose             Enable extra logging of requests/responses
      --force-insecure      Allow unverified TLS certificates when querying
                            service
      --custom-auth-token=DCOS_AUTH_TOKEN
                            Custom auth token to use when querying service
      --custom-dcos-url=DCOS_URI/DCOS_URL
                            Custom cluster URL to use when querying service
      --custom-cert-path=DCOS_CA_PATH/DCOS_CERT_PATH
                            Custom TLS CA certificate file to use when querying
                            service
      --name="hello-world"  Name of the service instance to query

Commands:
  help [<command>...]
    Show help.

  config list
    List IDs of all available configurations

  config show <config_id>
    Display a specified configuration

  config target
    Display the target configuration

  config target_id
    List ID of the target configuration

  describe
    View the package configuration for this DC/OS service

  endpoints [<name>]
    View client endpoints

  plan force-complete <plan> <phase> <step>
    Force complete a specific step in the provided phase

  plan force-restart <plan> [<phase>] [<step>]
    Restart a deploy plan, or specific step in the provided phase

  plan list
    Show all plans for this service

  plan pause [<plan>] [<phase>]
    Pause the deploy plan, or the plan with the provided name, or a specific
    phase in that plan with the provided name or UUID

  plan resume [<plan>] [<phase>]
    Resume the deploy plan, or the plan with the provided name, or a specific
    phase in that plan with the provided name or UUID

  plan start [<flags>] <plan>
    Start the plan with the provided name, with optional envvars to supply to
    task

  plan status [<flags>] [<plan>]
    Display the deploy plan or the plan with the provided name

  plan stop <plan>
    Stop the plan with the provided name

  pods list
    Display the list of known pod instances

  pods status [<pod>]
    Display the status for tasks in one pod or all pods

  pods info <pod>
    Display the full state information for tasks in a pod

  pods restart <pod>
    Restarts a given pod without moving it to a new agent

  pods replace <pod>
    Destroys a given pod and moves it to a new agent

  state framework_id
    Display the Mesos framework ID

  state properties
    List names of all custom properties

  state property <name>
    Display the content of a specified property

  state refresh_cache
    Refresh the state cache, used for debugging

  update force-complete <plan> <phase> <step>
    Force complete a specific step in the provided phase

  update force-restart [<plan>] [<phase>] [<step>]
    Restart a deploy plan, or specific step in the provided phase

  update pause [<plan>]
    Pause the deploy plan, or the plan with the provided name, or a specific
    phase in that plan with the provided name or UUID

  update resume [<plan>]
    Resume the deploy plan, or the plan with the provided name, or a specific
    phase in that plan with the provided name or UUID

  update status [<flags>] [<plan>]
    View status of the current update
```

### CODE REVIEW

**Comment by benclarkwood** [comment]
@ssk2, For update *, you shouldn't need to specify a plan.

Update will either be the update plan (if one is defined) or the deploy plan.

File: sdk/scheduler/src/main/java/com/mesosphere/sdk/api/PlansResource.java [hunk 1]
```diff
@@ -24,12 +24,12 @@
 @Path("/v1")
 public class PlansResource extends PrettyJsonResource {
 
-    static final Response ELEMENT_NOT_FOUND_RESPONSE = plainResponse("Element not found", Response.Status.NOT_FOUND);
     private static final StringMatcher ENVVAR_MATCHER = RegexMatcher.create("[A-Za-z_][A-Za-z0-9_]*");
 
     private final Logger logger = LoggerFactory.getLogger(getClass());
     private final Collection<PlanManager> planManagers;
 
+
```
Comment (ssk2) [inline sdk/scheduler/src/main/java/com/mesosphere/sdk/api/PlansResource.java]: Fixed.

File: cli/client/dcoscli.go [hunk 2]
```diff
@@ -11,42 +10,48 @@ import (
...
-		log.Fatalf("CLI configuration value %s (%s) is missing/unset. %s",
+		PrintMessageAndExit("CLI configuration value %s (%s) is missing/unset. %s",
 			name, description, errorInstruction)
 	}
 	return output
 }
 
+// OptionalCLIConfigValue retrieves the CLI configuration for name. If no value can
+// be retrieved, this returns an empty string.
 func OptionalCLIConfigValue(name string) string {
```
Comment (ssk2) [inline cli/client/dcoscli.go]: This is used in commands.go in the New method. I think it has to stay public for now.

File: cli/client/dcoscli.go [hunk 3]
```diff
@@ -11,42 +10,48 @@ import (
...
 		}
 		return string(outBytes), err
 	}
 	// trim any trailing newline (or any other whitespace) if present:
 	return strings.TrimSpace(string(outBytes)), nil
 }
 
+// RequiredCLIConfigValue retrieves the CLI configuration property for name. If no value can
+// be retrieved, this terminates with a fatal error.
 func RequiredCLIConfigValue(name string, description string, errorInstruction string) string {
```
Comment (ssk2) [inline cli/client/dcoscli.go]: Fixed.

File: cli/client/dcoscli.go [hunk 4]
```diff
@@ -11,42 +10,48 @@ import (
 
 // TODO(nick): Consider breaking this config retrieval out into a separate independent library?
 
+// RunCLICommand is used to run generic commands against the DC/OS CLI.
+// It attempts to run the `dcos` executable contained within the user's PATH.
 func RunCLICommand(arg ...string) (string, error) {
```
Comment (ssk2) [inline cli/client/dcoscli.go]: Fixed.

File: cli/client/response.go [hunk 5]
```diff
@@ -1,78 +1,63 @@
...
-		LogMessageAndExit("Failed to read response data from %s %s query: %s",
-			response.Request.Method, response.Request.URL, err)
+		return nil, err
 	}
-	return responseBytes
+	return responseBytes, nil
 }
 
+// UnmarshalJSON unmarshals a []byte of JSON into a map[string]interface{}
+// for easy handling of JSON responses that have variable field.
```
Comment (ssk2) [inline cli/client/response.go]: 1. Fixed. 2. It's used by various commands.

File: cli/client/response.go [hunk 6]
```diff
@@ -1,78 +1,63 @@
...
-}
+var customCheck responseCheck
 
-func printServiceNameErrorAndExit(response *http.Response) {
-	printResponseError(response)
-	LogMessage("- Did you provide the correct service name? Currently using '%s', specify a different name with '--name=<name>'.", config.ServiceName)
-	LogMessageAndExit("- Was the service recently installed or updated? It may still be initializing, wait a bit and try again.")
+// SetCustomResponseCheck sets a custom responseCheck that
+// can be used for more advanced handling of HTTP responses.
+func SetCustomResponseCheck(check responseCheck) {
```
Comment (ssk2) [inline cli/client/response.go]: Unfortunately this is used in the `commands` package too.

Comment (ssk2) [inline cli/client/response.go]: (Some commands require advanced handling of HTTP responses based on error codes and the message contents - e.g. the Plans API, and the Cosmos API.)

File: cli/client/cosmos.go [hunk 7]
```diff
@@ -68,16 +75,16 @@ func createCosmosHTTPJSONRequest(method, urlPath, jsonPayload string) *http.Requ
...
 	// Try to fetch the Cosmos URL from the system configuration
-	if len(config.CosmosUrl) == 0 {
-		config.CosmosUrl = OptionalCLIConfigValue("package.cosmos_url")
+	if len(config.CosmosURL) == 0 {
+		config.CosmosURL = OptionalCLIConfigValue("package.cosmos_url")
 	}
 
 	// Use Cosmos URL if we have it specified
-	if len(config.CosmosUrl) > 0 {
+	if len(config.CosmosURL) > 0 {
```
Comment (ssk2) [inline cli/client/cosmos.go]: I'm not sure I follow. The first conditional checks to see if we have one specified - if we haven't, then it'll attempt to read it from the DC/OS CLI's configuration. This is similar to how http.go does it. The reason we don't always set it is that the user may manually override it (in fact, we do this for unit testing).

The second conditional uses that URL if we have it specified - I can't see of a clean way to merge this into the previous conditional that wouldn't be more confusing than the current implemtnation.

File: cli/client/cosmos.go [hunk 8]
```diff
@@ -68,16 +75,16 @@ func createCosmosHTTPJSONRequest(method, urlPath, jsonPayload string) *http.Requ
 
 func createCosmosURL(urlPath string) *url.URL {
 	// Try to fetch the Cosmos URL from the system configuration
-	if len(config.CosmosUrl) == 0 {
-		config.CosmosUrl = OptionalCLIConfigValue("package.cosmos_url")
+	if len(config.CosmosURL) == 0 {
+		config.CosmosURL = OptionalCLIConfigValue("package.cosmos_url")
```
Comment (ssk2) [inline cli/client/cosmos.go]: Done. Just to clarify - did you mean "package.cosmos_url"?

File: cli/client/cosmos.go [hunk 9]
```diff
@@ -10,51 +10,58 @@ import (
...
+func parseCosmosHTTPErrorResponse(response *http.Response, body []byte) error {
+	responseJSON, err := UnmarshalJSON(body)
 	if err != nil {
-		printResponseErrorAndExit(response)
+		return createResponseError(response)
 	}
 	if errorType, present := responseJSON["type"]; present {
 		message := responseJSON["message"]
 		switch errorType {
 		case "MarathonAppNotFound":
```
Comment (ssk2) [inline cli/client/cosmos.go]: Done.

File: cli/client/cosmos.go [hunk 10]
```diff
@@ -10,51 +10,58 @@ import (
...
 
-func printBadVersionErrorAndExit(response *http.Response, data map[string]interface{}) {
-	requestedVersion := data["updateVersion"]
-	//TODO: this is probably an array?
-	validVersions := data["validVersions"]
-	printResponseError(response)
-	LogMessage("- Unable to update %s to requested version: %s", config.ServiceName, requestedVersion)
-	LogMessageAndExit("- Valid versions are: %s", validVersions)
+func createBadVersionError(response *http.Response, data map[string]interface{}) error {
+	requestedVersion, _ := GetValueFromJSON(data, "updateVersion")
```
Comment (ssk2) [inline cli/client/cosmos.go]: Good catch, fixed!

**Comment by ssk2** [comment]
@benclarkwood: addressed your review comments. Working on the default plan name logic - this might require some extra work.

**Comment by benclarkwood** [comment]
Apparently intellij didn't do a good job with Go, which is why I thought all that stuff could be package private.

Pending the default plan name handling I'm good with this.

**Comment by ssk2** [comment]
Tested package versions:
```
➜  cli git:(a323b42) ✗ dcos hello-world update package-versions
Current package version is: "1.1.6"
Package can be downgraded to: ["1.0.0","1.1.0","1.1.1","1.1.2","1.1.3","1.1.4","1.1.5","1.1.6"]
No valid package upgrade versions.
```

**Comment by ssk2** [comment]
@benclarkwood, updated.

**Comment by benclarkwood** [comment]
retest this please

What I believe to be a fix for this is here: https://github.com/mesosphere/dcos-commons/pull/1133

**Comment by benclarkwood** [comment]
retest this please

**Comment by benclarkwood** [comment]
retest this please

**Comment by benclarkwood** [comment]
retest this please (CCM)

**Comment by benclarkwood** [comment]
retest this please (CCM)

**Comment by ssk2** [comment]
retest this please (CCM)

**Comment by ssk2** [comment]
I am racing CI to see if this passes locally ;-)

**Comment by ssk2** [comment]
Looks like the hello-world tests timed out because CCM deleted the cluster before it was finished:
```

HTTP Connection Error
Traceback (most recent call last):
  File "/jenkins/workspace/dcos-commons/inf-pr/helloworld/python_deps/requests/packages/urllib3/connection.py", line 142, in _new_conn
    (self.host, self.port), self.timeout, **extra_kw)
  File "/jenkins/workspace/dcos-commons/inf-pr/helloworld/python_deps/requests/packages/urllib3/util/connection.py", line 67, in create_connection
    for res in socket.getaddrinfo(host, port, 0, socket.SOCK_STREAM):
  File "/usr/lib/python3.5/socket.py", line 732, in getaddrinfo
    for res in _socket.getaddrinfo(host, port, family, type, proto, flags):
socket.gaierror: [Errno -2] Name does not resolve
```

I'm tired of waiting for this CI to pass and have managed to run the tests successfully locally, so I'm merging this.

**Pull request merged**